csrc/preseg_passes/translate_repeat_to_expand.cpp

```diff
@@ -0,0 +1,194 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <preseg_passes/translate_repeat_to_expand.h>
+
+#include <ir/utils.h>
+#include <logical_domain_map.h>
+#include <ops/all_ops.h>
+
+#include <unordered_map>
+#include <vector>
+
+namespace nvfuser::preseg_passes {
+
+namespace {
+
+struct RepetitionInfo {
+  // Input tensor that is repeated
+  TensorView* input_tv = nullptr;
+  // Repeated logical ID of the input tensor
+  IterDomain* repeated_id = nullptr;
+  // Tensors fed into the concat op
+  std::vector<TensorView*> cat_inp_tvs;
+};
+
+// Translation algorithm overview:
+//
+// Step 1: Inspection. Traverses the given fusion and looks for a
+// sequence of ops that correspond to a repeatition. See
+// RepeatToExpandTranslator::inspect() for more details.
+//
+// Step 2: Apply the translation in a reverse topologial order. See
+// RepeatToExpandTranslator::translate() for more details.
+class RepeatToExpandTranslator {
+ public:
+  RepeatToExpandTranslator(Fusion* fusion) : fusion_(fusion) {}
+
+  void run() {
+    inspect();
+    translate();
+  }
+
+ private:
+  // Traverse through the fusion and gather all patterns of a pad
+  // followed by a concat. If a single concat op has multiple pad
+  // inputs that resize the same iter domain of the same input tensor,
+  // that must correspond to a repetition.
+  void inspect() {
+    const auto exprs = fusion_->exprs();
+
+    for (auto pad : ir_utils::filterByType<PadOp>(exprs)) {
```

Comment:
QQ: what's the reason that we choose to filter on `PadOp` here, instead of looking for `CatOp` and iterate through all its input?

It's just a nitpicking question, I think when we iterate through each `PadOp`, we could repetitively remove entries and adding things back into `repeat_info_map_`.

Reply:
I think I was initially also thinking about supporting concatenation without `CatOp`. I think I have seen a sequence of ops like a `PadOp` followed by an addition, maybe in some backward kernels. 

I dropped that since I don't find that pattern for repetitions. 
https://github.com/NVIDIA/Fuser/pull/3645/commits/4c1abda909fbf2d85786b13f3d1566a3490db567